carma-messenger-core/cpp_message/src/cpp_message.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (C) 2018-2020 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * CPP File containing Message method implementations
+ */
+
+#include "cpp_message.h"
+
+namespace cpp_message
+{
+
+    void Message::initialize()
+    {
+        nh_.reset(new ros::CARMANodeHandle());
+        pnh_.reset(new ros::CARMANodeHandle("~"));
+        // initialize pub/sub
+        outbound_binary_message_pub_ = nh_->advertise<cav_msgs::ByteArray>("outbound_binary_msg", 5);
+        inbound_binary_message_sub_ = nh_->subscribe("inbound_binary_msg", 5, &Message::inbound_binary_callback, this);
+        outbound_geofence_request_message_sub_ = nh_->subscribe("outbound_control_request", 5, &Message::outbound_control_request_callback, this);
+        inbound_geofence_request_message_pub_ = nh_->advertise<j2735_msgs::ControlRequest>("inbound_control_request", 5);
+    }
+
+    void Message::inbound_binary_callback(const cav_msgs::ByteArrayConstPtr& msg)
+    {
+        // only handle ControlRequest for now
+        if(msg->messageType == "ControlRequest") {
+            j2735_msgs::ControlRequest output;
+            // decode results
+            asn_dec_rval_t rval;
+	        MessageFrame_t* message = 0;
+            // copy from vector to array
+	        std::vector<uint8_t> array = msg->content;
+            auto len = array.size();
+            uint8_t buf[len];
+            for(auto i = 0; i < len; i++) {
+	            buf[i] = array[i];
+	        }
+            // use asn1c lib to decode
+	        rval = uper_decode(0, &asn_DEF_MessageFrame, (void **) &message, buf, len, 0, 0);
+
+            // if decode successed
+            if(rval.code == RC_OK) {
+                // convert version from char array to string
+                std::string version;
+                auto str_len = message->value.choice.TestMessage04.body.version.size;
```

Comment:
have you tried just doing
```version.append(message->value.choice.TestMessage04.body.version);```
to covert the string

Reply:
Nope. Does not work. The total allocated char array size is not equal to its actual size of the content.

Reply:
Ah interesting. This is fine then.